geo-types/src/vertex_string.rs

```diff
@@ -0,0 +1,674 @@
+            Cost::Haversine => "Haversine",
+            Cost::Customize(_) => "Customized_Cost_Function",
+        };
+
+        write!(f, "{}", name)
+    }
+}
+
+impl<T> PartialEq for Cost<T>
+where
+    T: Float,
+{
+    fn eq(&self, other: &Self) -> bool {
+        match (self, other) {
+            (Cost::Euclidean, Cost::Euclidean) => true,
+            (Cost::Haversine, Cost::Haversine) => true,
+            (Cost::Customize(func_self), Cost::Customize(func_other)) => {
+                std::ptr::eq(func_self, func_other)
+            },
+            _ => false
+        }
+    }
+}
+
+/// The struct representing the edge, aka graph relation between 2 neighboring and connected vertices.
+/// This struct is mostly used internally, though we're exposing APIs to retrieve fields in case
+/// applications would find them useful.
+#[derive(Debug)]
+#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
+pub struct GraphRelation {
+    // the neighbor vertex id
+    vertex_id: usize,
+
+    // the cost of the edge connecting the vertex and this neighbor
+    cost: f64,
+}
+
+impl GraphRelation {
+    /// Build a graph relationship between two neighboring vertices and connect them. The source
+    /// vertex is where the `GraphRelation` is added to, and the target vertex's id is saved into
+    /// the struct. The cost is calculated and passed in as the known value.
+    pub fn new_with_cost<T: Float>(index: usize, cost: T) -> Self {
+        GraphRelation {
+            vertex_id: index,
+            cost: cost.to_f64().unwrap_or(-1f64),
+        }
+    }
+
+    /// Get the vertex id of the other end of the edge
+    pub fn neighbor_id(&self) -> usize {
+        self.vertex_id
+    }
+
+    /// Get the cost of traveling this edge to the other end
+    pub fn edge_cost(&self) -> f64 {
+        self.cost
+    }
+}
+
+/// The struct representing the vertex in the `VertexString`.
+#[derive(Debug)]
+#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
+pub struct Vertex<T>
+where
+    T: Float
+{
+    // the node location
+    coordinate: Coordinate<T>,
+
+    // the vertex id
+    id: usize,
+
+    // vector of neighbor nodes that we can go to, stored in the form of (Destination, Distance)
+    vertices: Vec<GraphRelation>,
+
+    // the closest vertex to self
+    closest_neighbor: Option<(usize, f64)>,
+}
+
+impl<T> Vertex<T>
+where
+    T: Float
+{
+    /// Create a new vertex. Parameters:
+    /// - `coordinate`: the coordinate of the vertex
+    /// - `id`: the id of the vertex, which is also the index of the vertex in the VertexString's
+    ///         `vector` store.
+    pub fn new(coordinate: Coordinate<T>, id: usize) -> Self {
+        Vertex {
+            coordinate,
+            id,
+            vertices: Vec::with_capacity(DEFAULT_SIZE),
+            closest_neighbor: None,
+        }
+    }
+
+    /// Get the vertex id, which also serves as the vector index in the VertexString
+    pub fn get_id(&self) -> usize {
```

Comment:
getters in rust shouldn't have a `get_` prefix

https://rust-lang-nursery.github.io/api-guidelines/naming.html#getter-names-follow-rust-convention-c-getter

Reply:
Yep, I will remove the `get_` prefix